src/peft/tuners/oft/layer.py

```diff
@@ -14,106 +14,293 @@
 
 import math
 import warnings
-from typing import Any, List, Optional, Set, Tuple
+from typing import Any, List, Optional, Set, Union
 
 import torch
 import torch.nn as nn
+import torch.nn.functional as F
 
-from peft.tuners.lycoris_utils import LycorisLayer, check_adapters_to_merge
+from peft.tuners.tuners_utils import BaseTunerLayer, check_adapters_to_merge
 
 
-class OFTLayer(nn.Module, LycorisLayer):
+class MultiplicativeDropoutLayer(nn.Module):
+    """
+    Implements the multiplicative dropout layer for BOFT.
+    """
+
+    def __init__(self, p=0.0):
+        """
+        Initializes the multiplicative dropout layer.
+
+        Parameters:
+        p (float): The probability of dropping out a block. Defaults to 0.0.
+        """
+        super().__init__()
+        self.p = p
+
+    def forward(self, x):
+        """
+        Applies multiplicative dropout to the input tensor.
+
+        Parameters:
+        x (Tensor): The input tensor of shape (D, H, H), where `D` represents
+                    the number of OFT blocks, and `H` is the size of the square
+                    blocks along the last two dimensions, the block size in OFT.
+        """
+        if self.training:
+            # Ensure the last two dimensions are the same
+            if x.shape[-1] != x.shape[-2]:
+                raise ValueError("The last two dimensions of input should be the same!")
+
+            D, H, _ = x.shape
+
+            # If block share, skip the multiplicative dropout
+            if D == 1:
+                return x
+
+            # Create a mask with 1s for matrices to be replaced with identity and 0s otherwise
```

Comment:
Sentence doesn't quite make sense, could you please check again?

Reply:
updated.